File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        
        Map<Integer, Integer> map1 = new HashMap<>();
        Map<Integer, Integer> map2 = new HashMap<>();
        
        for (int top : topping) {
            map1.put(top, map1.getOrDefault(top, 0) + 1);
        }
        
        for (int top : topping) {
            map2.put(top, map2.getOrDefault(top, 0) + 1);
            
            if (map1.get(top) - 1 == 0) {
                map1.remove(top);
            } else {
                map1.put(top, map1.get(top) - 1);
            }
            
            if (map1.size() == map2.size()) {
                answer++;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter
def solution(topping):
    answer = 0
    
    dic = Counter(topping)
    setDic = set()
    
    for i in topping:
        dic[i] -= 1
        setDic.add(i)
        
        if dic[i] == 0:
            dic.pop(i)
        if len(dic) == len(setDic):
            answer += 1
    
    
    return answer